Added key features to TS (#419)
exercism/typescript: TAGS.md

# Tags

This document aims to provide reasoning why `config.json` contains the `"tags"` it contains.

> It uses as a basis the Javacript's [TAGS.md](https://github.com/exercism/javascript/blob/main/TAGS.md) because anything that Javascript does can be done with Typescript, being a transpiled language, and adds/modifies with the perks Typescript brings on top of it.

Reviewer: ```suggestion
It uses as a basis the JavaScript's [TAGS.md](https://github.com/exercism/javascript/blob/main/TAGS.md) because anything that JavaScript does can be done with Typescript, being a transpiled language, and adds/modifies with the perks TypeScript brings on top of it.
```

## Paradigms

- [x] `paradigm/declarative`: mostly popularised by libraries and frameworks such as React, Vue, etc.
- [x] `paradigm/functional`: there is a lot of support for functional programming, including various defactor libraries in the ecosystem providing functional programming patterns.

> - [x] `paradigm/imperative`: It retains Javascript's script and DOM's manipulation imperative programming, but the language design doesn't encourage it.

Reviewer: ```suggestion
- [x] `paradigm/imperative`: It retains JavaScript's script and DOM's manipulation imperative programming, but the language design doesn't encourage it.
```

> - [ ] `paradigm/logic`: whilst it is possible to write DSL or patterns that' allow for logic-based programming, JavaScript doesn't inheritly support it.

Reviewer: ```suggestion
- [ ] `paradigm/logic`: whilst it is possible to write DSL or patterns that' allow for logic-based programming, JavaScript doesn't inherently support it.
```

> - [x] `paradigm/object_oriented`: becomes a main feature of the language compared to Javascript, along with static type checking and a rich type system.

Reviewer: ```suggestion
- [x] `paradigm/object_oriented`: becomes a main feature of the language compared to JavaScript, along with static type checking and a rich type system.
```

## Typing

- [x] `typing/static`: will check types at compile time for any part of the code which has its types stated.

> - [x] `typing/dynamic`: done at runtime for any data that uses the `any` keyword.

Author: Changed the wording to `at runtime all typescript type information is stripped. Runtime checking can also be manually achieved as well`. How is that?

> - [x] `typing/dynamic`: done at runtime for any data that uses the `any` keyword.

Reviewer: There is a 100 character limit for these

> - [x] `typing/dynamic`: done at runtime for any data that uses the `any` keyword.

Reviewer: Sure!

- [x] `typing/strong`: Typescript adds a rich type system, that is check at compile time if used.

> - [x] `typing/weak`: Allows for use of weak typing using the `any` keyword.

Reviewer: ```suggestion
- [x] `typing/weak`: TypeScript allows for explicit (strong) typing, inferred (weak) typing, deferred (computed / conditional) typing, opt-out (using `any`) of typing, and forced run-time (using `unknown`) typing.
```

## Execution mode

> - [x] `execution_mode/compiled`: Typescript transpilation compiles the code to Javascript

Reviewer: ```suggestion
- [x] `execution_mode/compiled`: TypeScript transpilation (transpilation via [`tsc`](https://www.typescriptlang.org/), or stripping types and then compiling using [`@babel/preset-typescript`](https://babeljs.io/docs/en/babel-preset-typescript)) compiles the code to a certain version of JavaScript. This includes in-memory compilation using tools such as [`ts-node`](https://github.com/TypeStrong/ts-node), however due to the in-line nature and REPL like support, it might be considered interpreted.
```

> - [ ] `execution_mode/interpreted`: Although is transpiled to Javascript (which is interpreted), Typescript code is never directly interpreted.

Reviewer: ```suggestion
- [x] `execution_mode/interpreted`: TypeScript _can_ be interpreted Just-In-Time by stripping types, for example using [`@babel/preset-typescript`](https://babeljs.io/docs/en/babel-preset-typescript). 
```
